File: frontend/js/views/rent_increase_notice.js
// Rent increase notice generator — new rent (percent or flat), change, and the
// effective date from the service date + notice period, via
// /calc/rent-increase-notice. Previews live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '%');
let LAST_DOC = null;

export async function renderRentIncreaseNotice(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.rinc.h1.title">// RENT INCREASE NOTICE</span></h1>
        <p class="muted small" data-i18n="view.rinc.hint.intro">
            The written notice a landlord serves before raising rent on a periodic tenancy. Enter the
            increase as a percent or a flat amount; it computes the new rent, the dollar and percent
            change, and the effective date from the service date plus the required notice period. Many
            states set the minimum notice (and sometimes a cap) by the size of the increase. Drafting
            aid, not legal advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.rinc.h2.inputs">Notice details</h2>
            <form id="rinc-form" class="inline-form">
                <label><span data-i18n="view.rinc.label.state">State / jurisdiction</span>
                    <input type="text" name="state" value="California" required></label>
                <label><span data-i18n="view.rinc.label.landlord_name">Landlord name</span>
                    <input type="text" name="landlord_name" value=""></label>
                <label><span data-i18n="view.rinc.label.landlord_address">Landlord address</span>
                    <input type="text" name="landlord_address" value=""></label>
                <label><span data-i18n="view.rinc.label.landlord_phone">Landlord phone</span>
                    <input type="text" name="landlord_phone" value=""></label>
                <label><span data-i18n="view.rinc.label.tenant_name">Tenant name</span>
                    <input type="text" name="tenant_name" value=""></label>
                <label><span data-i18n="view.rinc.label.premises">Premises address</span>
                    <input type="text" name="premises_address" value=""></label>
                <label><span data-i18n="view.rinc.label.current_rent">Current rent ($)</span>
                    <input type="number" step="0.01" min="0" name="current_rent_usd" value="1500" required></label>
                <label><span data-i18n="view.rinc.label.type">Increase by</span>
                    <select name="increase_type" id="rinc-type">
                        <option value="percent" data-i18n="view.rinc.type.percent">Percent (%)</option>
                        <option value="amount" data-i18n="view.rinc.type.amount">Flat amount ($)</option>
                    </select></label>
                <label><span data-i18n="view.rinc.label.value">Increase value</span>
                    <input type="number" step="0.01" min="0" name="increase_value" value="5" required></label>
                <label><span data-i18n="view.rinc.label.served">Date served</span>
                    <input type="date" name="served_date" value="2026-06-01" required></label>
                <label><span data-i18n="view.rinc.label.notice_days">Notice period (days)</span>
                    <input type="number" step="1" min="1" name="notice_days" value="60" required></label>
                <label><span data-i18n="view.rinc.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.rinc.ph.statute'))}"></label>
            </form>
        </div>
        <div id="rinc-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#rinc-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            landlord_name: (fd.get('landlord_name') || '').trim(),
            landlord_address: (fd.get('landlord_address') || '').trim(),
            landlord_phone: (fd.get('landlord_phone') || '').trim(),
            tenant_name: (fd.get('tenant_name') || '').trim(),
            premises_address: (fd.get('premises_address') || '').trim(),
            current_rent_usd: Number(fd.get('current_rent_usd')) || 0,
            increase_type: fd.get('increase_type'),
            increase_value: Number(fd.get('increase_value')) || 0,
            served_date: fd.get('served_date'),
            notice_days: Math.round(Number(fd.get('notice_days')) || 0),
            state: (fd.get('state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcRentIncreaseNotice(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.rinc.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#rinc-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.rinc.card.new_rent">New rent</div>
                    <div class="value">${money(doc.new_rent_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.rinc.card.increase">Increase</div>
                    <div class="value">${money(doc.increase_amount_usd)} · ${pct(doc.increase_pct)}</div></div>
                <div class="card neg"><div class="label" data-i18n="view.rinc.card.effective">Effective</div>
                    <div class="value">${esc(doc.effective_date || '—')}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="rinc-copy" type="button" data-i18n="view.rinc.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="rinc-download" type="button" data-i18n="view.rinc.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#rinc-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.rinc.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.rinc.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#rinc-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'rent-increase-notice.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/rent_increase_notice.rs
//! Rent increase notice — the written notice a landlord must serve before
//! raising rent on a periodic tenancy. It computes the new rent (by percent or
//! flat amount), the dollar and percent change, and the effective date from the
//! service date plus the required notice period, then assembles the notice.
//! Drafting aid, not legal advice — many states set the minimum notice (and
//! sometimes a cap) by the size of the increase.

use chrono::{Duration, NaiveDate};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum IncreaseType {
    Percent,
    Amount,
}

#[derive(Debug, Clone, Deserialize)]
pub struct RentIncreaseInput {
    pub landlord_name: String,
    pub landlord_address: String,
    #[serde(default)]
    pub landlord_phone: String,
    pub tenant_name: String,
    pub premises_address: String,
    pub current_rent_usd: f64,
    pub increase_type: IncreaseType,
    /// Percent (when Percent) or dollars (when Amount).
    pub increase_value: f64,
    /// Date the notice is served (YYYY-MM-DD).
    pub served_date: String,
    /// Days of notice before the increase takes effect (per state minimum).
    pub notice_days: i64,
    pub state: String,
    #[serde(default)]
    pub statute_citation: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct RentIncreaseNotice {
    pub title: String,
    pub current_rent_usd: f64,
    pub new_rent_usd: f64,
    pub increase_amount_usd: f64,
    pub increase_pct: f64,
    pub effective_date: String,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &RentIncreaseInput) -> RentIncreaseNotice {
    let new_rent = match i.increase_type {
        IncreaseType::Percent => i.current_rent_usd * (1.0 + i.increase_value / 100.0),
        IncreaseType::Amount => i.current_rent_usd + i.increase_value,
    };
    let increase_amount = new_rent - i.current_rent_usd;
    let increase_pct = if i.current_rent_usd != 0.0 {
        increase_amount / i.current_rent_usd * 100.0
    } else {
        0.0
    };

    let effective_date = NaiveDate::parse_from_str(&i.served_date, "%Y-%m-%d")
        .map(|d| (d + Duration::days(i.notice_days)).format("%Y-%m-%d").to_string())
        .unwrap_or_default();

    let citation = i.statute_citation.trim();
    let pursuant = if citation.is_empty() {
        format!(
            "This notice is given pursuant to the landlord-tenant law of the State of {}.",
            i.state
        )
    } else {
        format!(
            "This notice is given pursuant to the landlord-tenant law of the State of {} ({}).",
            i.state, citation
        )
    };

    let clauses = vec![
        DocClause { heading: "To".into(), body: i.tenant_name.clone() },
        DocClause {
            heading: "1. Current Rent".into(),
            body: format!(
                "The current monthly rent for the premises located at {} is {}.",
                i.premises_address,
                money(i.current_rent_usd)
            ),
        },
        DocClause {
            heading: "2. New Rent and Effective Date".into(),
            body: format!(
                "Effective {}, the monthly rent will increase by {} ({:.2}%) to {}. All other terms of your tenancy remain unchanged. If you do not wish to continue the tenancy at the new rent, you must vacate on or before the effective date in accordance with your rental agreement and applicable law.",
                effective_date,
                money(increase_amount),
                increase_pct,
                money(new_rent)
            ),
        },
        DocClause {
            heading: "3. Notice Period".into(),
            body: format!(
                "This notice is served {} and provides {} days' notice before the increase takes effect.",
                i.served_date, i.notice_days
            ),
        },
        DocClause { heading: "4. Governing Law".into(), body: pursuant },
        DocClause {
            heading: "Signature".into(),
            body: format!(
                "Signature of owner of premises or agent: ____________________  Date: {}\n{}\n{}{}",
                i.served_date,
                i.landlord_name,
                i.landlord_address,
                if i.landlord_phone.is_empty() {
                    String::new()
                } else {
                    format!("\nTelephone: {}", i.landlord_phone)
                }
            ),
        },
    ];

    RentIncreaseNotice {
        title: "Notice of Rent Increase".into(),
        current_rent_usd: i.current_rent_usd,
        new_rent_usd: new_rent,
        increase_amount_usd: increase_amount,
        increase_pct,
        effective_date,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-6
    }

    fn base() -> RentIncreaseInput {
        RentIncreaseInput {
            landlord_name: "Acme Property Mgmt".into(),
            landlord_address: "1 Main St".into(),
            landlord_phone: String::new(),
            tenant_name: "Jane Doe".into(),
            premises_address: "42 Rental Rd".into(),
            current_rent_usd: 1500.0,
            increase_type: IncreaseType::Percent,
            increase_value: 5.0,
            served_date: "2026-06-01".into(),
            notice_days: 60,
            state: "California".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn percent_increase() {
        let n = generate(&base());
        assert!(close(n.new_rent_usd, 1575.0));
        assert!(close(n.increase_amount_usd, 75.0));
        assert!(close(n.increase_pct, 5.0));
    }

    #[test]
    fn amount_increase_back_computes_percent() {
        let n = generate(&RentIncreaseInput {
            increase_type: IncreaseType::Amount,
            increase_value: 100.0,
            ..base()
        });
        assert!(close(n.new_rent_usd, 1600.0));
        assert!(close(n.increase_amount_usd, 100.0));
        assert!((n.increase_pct - 6.6667).abs() < 1e-3);
    }

    #[test]
    fn effective_date_is_served_plus_notice() {
        // 2026-06-01 + 60 days = 2026-07-31.
        assert_eq!(generate(&base()).effective_date, "2026-07-31");
    }

    #[test]
    fn zero_current_rent_no_divide_panic() {
        let n = generate(&RentIncreaseInput { current_rent_usd: 0.0, ..base() });
        assert!(close(n.increase_pct, 0.0));
    }

    #[test]
    fn statute_citation_echoed() {
        let n = generate(&RentIncreaseInput {
            statute_citation: "Cal. Civ. Code § 827".into(),
            ..base()
        });
        assert_eq!(n.statutory_citation, "Cal. Civ. Code § 827");
        assert!(n.clauses.iter().any(|c| c.body.contains("Cal. Civ. Code § 827")));
    }

    #[test]
    fn new_rent_appears_in_clause() {
        let n = generate(&base());
        let c = n.clauses.iter().find(|c| c.heading.contains("New Rent")).unwrap();
        assert!(c.body.contains("$1575.00"));
    }

    #[test]
    fn bad_date_yields_empty_effective() {
        let n = generate(&RentIncreaseInput { served_date: "x".into(), ..base() });
        assert_eq!(n.effective_date, "");
    }
}
